packages/docusaurus/src/client/PendingNavigation.tsx

```diff
@@ -61,24 +61,37 @@ class PendingNavigation extends React.Component<Props, State> {
     // Load data while the old screen remains.
     preload(nextLocation.pathname)
       .then(() => {
-        clientLifecyclesDispatcher.onRouteUpdate({
-          previousLocation: this.previousLocation,
-          location: nextLocation,
-        });
-        this.setState({nextRouteHasLoaded: true}, this.stopProgressBar);
-        const {hash} = nextLocation;
-        if (!hash) {
-          window.scrollTo(0, 0);
-        } else {
-          const id = decodeURIComponent(hash.substring(1));
-          const element = document.getElementById(id);
-          element?.scrollIntoView();
+        if (this.previousLocation?.pathname !== nextLocation.pathname) {
+          clientLifecyclesDispatcher.onRouteUpdate({
+            previousLocation: this.previousLocation,
+            location: nextLocation,
+          });
         }
+        this.setState({nextRouteHasLoaded: true}, this.stopProgressBar);
       })
       .catch((e) => console.warn(e));
     return false;
   }
 
+  override componentDidUpdate(): void {
+    if (this.previousLocation !== this.props.location) {
+      const {hash} = this.props.location;
+      if (!hash) {
+        window.scrollTo(0, 0);
+      } else {
+        const id = decodeURIComponent(hash.substring(1));
+        const element = document.getElementById(id);
+        element?.scrollIntoView();
+      }
+    }
```

Comment:
Moving this scrolling to `componentDidUpdate` fixes the long-standing issue that hash links don't work cross-page in development. Now, when calling `scrollIntoView`, the anchored DOM always exists even when it's the first render.